physics/cooling/include/cooling/cooler_impl.hpp

```diff
@@ -0,0 +1,389 @@
+
+        grackle_fields.makeGrackleFieldsFromData(rho, u, particle, len);
+        auto ret_value = local_calculate_pressure(&global_values.data, &global_values.rates, &global_values.units,
+                                                  &grackle_fields.data, p);
+        if (ret_value == 0) { throw std::runtime_error("Grackle: local_calculate_pressure"); }
+    }
+
+    void cooling_time_arr(T* rho, T* u, const ParticleType& particle, T* ct, const size_t len)
+    {
+        static_assert(std::is_same_v<T, gr_float>);
+
+        cooler_field_data_arr<T, BlockSize> grackle_fields;
+
+        grackle_fields.makeGrackleFieldsFromData(rho, u, particle, len);
+
+        auto ret_value = local_calculate_cooling_time(&global_values.data, &global_values.rates, &global_values.units,
+                                                      &grackle_fields.data, ct);
+        if (ret_value == 0) { throw std::runtime_error("Grackle: local_calculate_cooling_time"); }
+    }
+
+    void adiabatic_index_arr(T* rho, T* u, const ParticleType& particle, T* gamma, const size_t len)
+    {
+        static_assert(std::is_same_v<T, gr_float>);
+
+        cooler_field_data_arr<T, BlockSize> grackle_fields;
+
+        grackle_fields.makeGrackleFieldsFromData(rho, u, particle, len);
+
+        auto ret_value = local_calculate_gamma(&global_values.data, &global_values.rates, &global_values.units,
+                                               &grackle_fields.data, gamma);
+        if (ret_value == 0) { throw std::runtime_error("Grackle: local_calculate_gamma"); }
+    }
+
+    //! @brief for_each field a in A: a[:] = op(a[:], b[:])
+    template<class Tuple, typename Trho, class Op>
+    void forEachTransform(Tuple&& A, const Trho* b, size_t first, size_t last, Op&& op)
+    {
+        util::for_each_tuple([b, first, last, &op](auto* a)
+                             { std::transform(a + first, a + last, b + first, a + first, op); }, A);
+    }
+
+    template<typename Trho, typename Tu>
+    void cool_particles(const T dt, const Trho* rho, const Tu* u, const ParticleType& particle, Tu* du,
+                        const size_t first, const size_t last)
+    {
+        const task<BlockSize> t(first, last);
+        const auto            compute_du = [&](const auto& u_block, const block& b)
+        {
+            for (size_t i = 0; i < b.len; i++)
+            {
+                du[i + b.first] += (u_block[i] - u[i + b.first]) / dt;
+            }
+        };
+
+#pragma omp parallel for
+        for (size_t i = 0; i < t.n_bins; i++)
+        {
+            std::array<T, BlockSize> rho_copy;
+            std::array<T, BlockSize> u_copy;
+
+            const block b(i, t);
+            copyToBlock(rho, rho_copy, b);
+            copyToBlock(u, u_copy, b);
+
+            //forEachTransform(util::get<Fractions, CoolingFields>(particle), rho, b.first, b.first + b.len,
+            //                 std::multiplies<>{});
+            convertFracToDens(particle, rho, b);
```

Comment:
I can "forget" to do the conversion from fractions to densities, either here or below by commenting the conversion, and the grackle cooling tests still pass!